src/Juvix/Compiler/Core/Extra.hs

```diff
@@ -98,6 +102,44 @@ captureFreeVars fv
             | Just v <- s ^. at (u - k) -> NVar (Var i (v + k))
           m -> m
 
+-- captures all free variables of a node. It also returns the list of captured
+-- variables in left-to-right order: if snd is of the form λxλy... then fst is
+-- [x, y]
+captureFreeVarsCtx :: BinderList Binder -> Node -> ([(Var, Binder)], Node)
+captureFreeVarsCtx bl n =
+  let assocs = freeVarsCtx bl n
+   in (assocs, captureFreeVars (map (first (^. varIndex)) assocs) n)
+
+freeVarsCtx' :: BinderList Binder -> Node -> [Var]
+freeVarsCtx' bl = map fst . freeVarsCtx bl
+
+-- | the output list does not contain repeated elements and is sorted by variable index.
+-- TODO it can probably be fore efficient
+freeVarsCtx :: BinderList Binder -> Node -> [(Var, Binder)]
```

Comment:
What are the free variables in the returned Binders relative to? It's not apparent from the code.

Reply:
I have added a clarification

Reply:
What I meant is that e.g. if you have the context [x : A$1, y : A$0, A : Type] and you close x then the result list is [x : A$0, A : Type]? At some point you need to shift the index of A in the type of x and I don't see where it happens.

Reply:
the shifting happens in `captureFreeVars`